File: src/sorting/select_mom.py
# 最壞情況線性 SELECT（CLRS §9.3 中位數之中位數；改寫自官方 order_statistics.py 精簡）
def select(A, i):
    """回傳 A 中第 i 小（1-based），最壞情況 O(n)。"""
    if len(A) <= 5:                       # 基底：小陣列直接排
        return sorted(A)[i - 1]
    # ① 5 個一組、各組取中位數
    medians = [sorted(A[j:j + 5])[len(A[j:j + 5]) // 2]
               for j in range(0, len(A), 5)]
    # ② 遞迴找「中位數們的中位數」當 pivot
    x = select(medians, (len(medians) + 1) // 2)
    # ③ 以 x 三路分割（保證兩側都至少甩掉約 3n/10 個元素）
    lo = [a for a in A if a < x]
    eq = [a for a in A if a == x]
    hi = [a for a in A if a > x]
    if i <= len(lo):
        return select(lo, i)
    if i <= len(lo) + len(eq):
        return x
    return select(hi, i - len(lo) - len(eq))


A = [3, 2, 9, 0, 7, 5, 4, 8, 6, 1, 12, 11, 10, 14, 13]
print("第 8 小（中位數）=", select(A, 8))
print("第 1 小 =", select(A, 1), "，第 15 小 =", select(A, 15))

File: src/sorting/select_mom.cpp
// 最壞情況線性 SELECT（CLRS §9.3 中位數之中位數）
#include <algorithm>
#include <iostream>
#include <vector>

int select(std::vector<int> A, int i) {          // 第 i 小（1-based），最壞 O(n)
    if (A.size() <= 5) {
        std::sort(A.begin(), A.end());
        return A[i - 1];
    }
    std::vector<int> medians;                    // ① 5 個一組取各組中位數
    for (std::size_t j = 0; j < A.size(); j += 5) {
        auto e = std::min(j + 5, A.size());
        std::vector<int> g(A.begin() + j, A.begin() + e);
        std::sort(g.begin(), g.end());
        medians.push_back(g[g.size() / 2]);
    }
    int x = select(medians, (medians.size() + 1) / 2);   // ② 中位數們的中位數
    std::vector<int> lo, eq, hi;                 // ③ 三路分割
    for (int a : A) (a < x ? lo : a == x ? eq : hi).push_back(a);
    if (i <= (int)lo.size()) return select(lo, i);
    if (i <= (int)(lo.size() + eq.size())) return x;
    return select(hi, i - lo.size() - eq.size());
}

int main() {
    std::vector<int> A = {3, 2, 9, 0, 7, 5, 4, 8, 6, 1, 12, 11, 10, 14, 13};
    std::cout << "第 8 小（中位數）= " << select(A, 8) << "\n";
    std::cout << "第 1 小 = " << select(A, 1)
              << "，第 15 小 = " << select(A, 15) << "\n";
    return 0;
}
